src/Corax/IndexWriter.cs

```diff
@@ -2009,13 +2004,20 @@ public void InsertTextualField()
                                 sortedTerms[idx], postListIds[idx],
                                 keys[idx].ContainerId, pageOffsets[idx]);
 
-                            processed++;
                             // if the tree structure changed, the bulk insert details are wrong
                             // and will need to restart the operation with a new BulkUpdateStart
                             if (treeChanged.Changed)
                             {
                                 // next time, we start from the _next_ key, not the current one
                                 idx++;
+                                for (int j = idx; j < read; j++)
+                                {
+                                    // Reset the known container id, since we modified the tree structure.
+                                    // The issue is that we may have a term id that was remembered by a separator key
+                                    // and we'll lose that after a page merge, so we'll have a reference to a deleted key
+                                    // see: RavenDB-21272
+                                    keys[j].ContainerId = -1;
+                                }
                                 break;
```

Comment:
That's nasty.

Reply:
That is actually going to be fairly rare, I think.
You need to have a page split, and we then need to reset the _rest_ of the values. 
On average, would happen rarely, and then only for half the results in a single page.

Assume common scenario is < 1000 items.